modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/service/impl/KaleoDefinitionVersionLocalServiceImpl.java

```diff
@@ -360,62 +361,76 @@ public int getLatestKaleoDefinitionVersionsCount(
 		return kaleoDefinitionVersionIds.size();
 	}
 
-	private void _addKeywordsCriterion(
-		DynamicQuery dynamicQuery, String keywords) {
-
-		if (Validator.isNull(keywords)) {
-			return;
-		}
-
-		Junction junction = RestrictionsFactoryUtil.disjunction();
-
-		for (String keyword : _customSQL.keywords(keywords)) {
-			junction.add(RestrictionsFactoryUtil.ilike("description", keyword));
-			junction.add(RestrictionsFactoryUtil.ilike("name", keyword));
-			junction.add(RestrictionsFactoryUtil.ilike("title", keyword));
-		}
-
-		dynamicQuery.add(junction);
-	}
-
-	private void _addStatusCriterion(DynamicQuery dynamicQuery, int status) {
-		if (status != WorkflowConstants.STATUS_ANY) {
-			Junction junction = RestrictionsFactoryUtil.disjunction();
-
-			junction.add(RestrictionsFactoryUtil.eq("status", status));
-
-			dynamicQuery.add(junction);
-		}
-	}
-
 	private List<Long> _getKaleoDefinitionVersionIds(
 		long companyId, String keywords, int status) {
 
-		List<Long> kaleoDefinitionVersionIds = new ArrayList<>();
-
-		DynamicQuery dynamicQuery = DynamicQueryFactoryUtil.forClass(
-			KaleoDefinitionVersion.class, getClassLoader());
-
-		Property companyIdProperty = PropertyFactoryUtil.forName("companyId");
-
-		dynamicQuery.add(companyIdProperty.eq(companyId));
+		KaleoDefinitionVersionTable aliasKaleoDefinitionVersionTable =
+			KaleoDefinitionVersionTable.INSTANCE.as(
+				"aliasKaleoDefinitionVersionTable");
+
+		DSLQuery dslQuery = DSLQueryFactoryUtil.select(
+			aliasKaleoDefinitionVersionTable.kaleoDefinitionVersionId
+		).from(
+			aliasKaleoDefinitionVersionTable
+		).where(
+			aliasKaleoDefinitionVersionTable.companyId.eq(
+				companyId
+			).and(
+				() -> {
+					if (Validator.isNull(keywords)) {
+						return null;
+					}
+
+					Predicate predicate = null;
+
+					for (String keyword : _customSQL.keywords(keywords)) {
+						predicate =
+							aliasKaleoDefinitionVersionTable.description.like(
+								keyword
+							).or(
+								aliasKaleoDefinitionVersionTable.name.like(
+									keyword)
+							).or(
+								aliasKaleoDefinitionVersionTable.title.like(
+									keyword)
+							);
+					}
+
+					return predicate.withParentheses();
+				}
+			).and(
+				() -> {
+					if (status == WorkflowConstants.STATUS_ANY) {
+						return null;
+					}
+
+					return aliasKaleoDefinitionVersionTable.status.eq(status);
+				}
+			).and(
+				aliasKaleoDefinitionVersionTable.version.in(
+					DSLQueryFactoryUtil.select(
+						DSLFunctionFactoryUtil.max(
+							DSLFunctionFactoryUtil.castLong(
+								KaleoDefinitionVersionTable.INSTANCE.version)
+						).as(
+							"latestVersion"
+						)
+					).from(
+						KaleoDefinitionVersionTable.INSTANCE
+					).where(
+						KaleoDefinitionVersionTable.INSTANCE.name.eq(
```

Comment:
Hi @MarinhoFeliphe , should we consider the `companyId` on this inner query?

Reply:
Yes! good catch!